drivers/platform/stm32/stm32_timer.c

```diff
@@ -0,0 +1,343 @@
+
+	switch (device_id) {
+#if defined(TIM1)
+	case 1:
+		*base = TIM1;
+		*clk_freq = apb2_freq;
+		break;
+#endif
+#if defined(TIM2)
+	case 2:
+		*base = TIM2;
+		break;
+#endif
+#if defined(TIM3)
+	case 3:
+		*base = TIM3;
+		break;
+#endif
+		break;
+#if defined(TIM4)
+	case 4:
+		*base = TIM4;
+		break;
+#endif
+#if defined(TIM5)
+	case 5:
+		*base = TIM5;
+		break;
+#endif
+#if defined(TIM6)
+	case 6:
+		*base = TIM6;
+		break;
+#endif
+#if defined(TIM7)
+	case 7:
+		*base = TIM7;
+		break;
+#endif
+#if defined(TIM8)
+	case 8:
+		*base = TIM8;
+		*clk_freq = apb2_freq;
+		break;
+#endif
+#if defined(TIM9)
+	case 9:
+		*base = TIM9;
+		break;
+#endif
+#if defined(TIM10)
+	case 10:
+		*base = TIM10;
+		break;
+#endif
+#if defined(TIM11)
+	case 11:
+		*base = TIM11;
+		break;
+#endif
+#if defined(TIM12)
+	case 12:
+		*base = TIM12;
+		break;
+#endif
+#if defined(TIM13)
+	case 13:
+		*base = TIM13;
+		break;
+#endif
+#if defined(TIM14)
+	case 14:
+		*base = TIM14;
+		break;
+#endif
+	default:
+		return -EINVAL;
+	};
+
+	return 0;
+}
+
+/**
+ * @brief Initialize the timer peripheral.
+ * @param desc  - The timer descriptor.
+ * @param param - The structure that contains the timer parameters.
+ * @return ret  - Result of the init procedure.
+ */
+int32_t stm32_timer_init(struct no_os_timer_desc **desc,
+			 const struct no_os_timer_init_param *param)
+{
+	int32_t ret;
+	struct no_os_timer_desc	*no_os_desc;
+	struct stm32_timer_desc *stm_desc;
+	uint32_t src_freq;
+	TIM_TypeDef *base = NULL;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	no_os_desc = (struct no_os_timer_desc *)calloc(1, sizeof(*no_os_desc));
+	if (!no_os_desc)
+		return -ENOMEM;
+
+	stm_desc = param->extra;
+
+	ret = get_timer_base(param->id, &base, &src_freq);
+	if (ret)
+		goto error;
+	/* Make sure the same id is used as the selected instance */
+	if (base != stm_desc->htimer->Instance) {
+		ret = -EINVAL;
+		goto error;
+	}
+
+	/* Make sure the selected frequency is in range */
+	if (src_freq < param->freq_hz) {
+		ret = -EINVAL;
+		goto error;
+	}
+
+	/* Overwrite generated values with given values */
+	stm_desc->htimer->Init.Prescaler = src_freq/param->freq_hz;
+	stm_desc->htimer->Init.Period = param->load_value;
+
+	ret = HAL_TIM_Base_Init(stm_desc->htimer);
+	if (ret != HAL_OK) {
+		ret = -EIO;
+		goto error;
+	}
+
+	/* Copy settings to device descriptor */
+	no_os_desc->extra = stm_desc->htimer;
+	no_os_desc->id = param->id;
+	no_os_desc->freq_hz = param->freq_hz;
+	no_os_desc->load_value = param->load_value;
+	*desc = no_os_desc;
+
+	return 0;
+
+error:
+	free(no_os_desc);
+	return ret;
+}
+
+/**
+ * @brief Free the resources allocated by no_os_timer_init().
+ * @param desc - The timer descriptor.
+ *
+ * @return ret - Result of the remove procedure.
+ */
+int32_t stm32_timer_remove(struct no_os_timer_desc *desc)
+{
+	if (!desc)
+		return -EINVAL;
+
+	if (!HAL_TIM_Base_DeInit(desc->extra)) {
+		free(desc);
+		return 0;
+	}
+
+	return -1;
```

Comment:
I think we need to do a better error reporting here, either:
- return -EIO
- or do some translation between STM32 HAL error codes and no-OS error codes

Reply:
Now -EIO is being used. Unfortunately STM32 HAL in this case returns either HAL_OK or HAL_ERROR so there is no useful information to be returned in case of an error (just the fact that an error occurred).